### PR TITLE
Issue 177 batch render calls

### PR DESCRIPTION
**Pull request by Derek52**

### CODE REVIEW

File: src/main/kotlin/kweb/Kweb.kt [hunk 1]
```diff
@@ -134,7 +134,7 @@ class Kweb private constructor(
         }
     }
 
-    private val clientState = CacheBuilder.newBuilder()
+    val clientState = CacheBuilder.newBuilder()
```
Comment (sanity) [inline src/main/kotlin/kweb/Kweb.kt:137]: Does this need to be public?

File: src/main/kotlin/kweb/Kweb.kt [hunk 2]
```diff
@@ -169,45 +169,39 @@ class Kweb private constructor(
...
+        val wsClientData = clientState.getIfPresent(sessionId)
+                ?: error("Client id $sessionId not found")
         wsClientData.lastModified = Instant.now()
         val debugToken: String? = if(!debug) null else {
             val dt = abs(random.nextLong()).toString(16)
             wsClientData.debugTokens[dt] = DebugInfo(javascript, "executing", Throwable())
             dt
         }
-        server2ClientMessage.debugToken = debugToken
+        funcCall.debugToken = debugToken
```
Comment (sanity) [inline src/main/kotlin/kweb/Kweb.kt:181]: Not necessary to fix, but might be good to add a TODO to figure out how to make debugToken immutable.

File: src/main/kotlin/kweb/Kweb.kt [hunk 3]
```diff
@@ -169,45 +169,39 @@ class Kweb private constructor(
...
         val outboundMessageCatcher = outboundMessageCatcher.get()
         if (outboundMessageCatcher == null) {
-            wsClientData.send(server2ClientMessage)
+            wsClientData.send(Server2ClientMessage(sessionId, funcCall))
         } else {
-            logger.debug("Temporarily storing message for ${server2ClientMessage.yourId} in threadlocal outboundMessageCatcher")
-            val jsFunction = JsFunction(server2ClientMessage.jsId!!, server2ClientMessage.arguments!!)
+            logger.debug("Temporarily storing message for $sessionId in threadlocal outboundMessageCatcher")
+            val jsFunction = JsFunction(funcCall.jsId!!, funcCall.arguments)
```
Comment (sanity) [inline src/main/kotlin/kweb/Kweb.kt:187]: Are `JsFunction` and `FuncCall` similar enough that we could use the same class for both?

File: src/main/kotlin/kweb/Kweb.kt [hunk 4]
```diff
@@ -169,45 +169,39 @@ class Kweb private constructor(
...
             outboundMessageCatcher.add(jsFunction)
             //Setting `shouldExecute` to false tells the server not to add this jsFunction to the client's cache,
             //but to not actually run the code. This is used to pre-cache functions on initial page render.
-            server2ClientMessage.shouldExecute = false
-            wsClientData.send(server2ClientMessage)
+            funcCall.shouldExecute = false
```
Comment (sanity) [inline src/main/kotlin/kweb/Kweb.kt:191]: Maybe add a TODO: to figure how how this can be immutable.

File: src/main/kotlin/kweb/Kweb.kt [hunk 5]
```diff
@@ -169,45 +169,39 @@ class Kweb private constructor(
...
         wsClientData.lastModified = Instant.now()
-        val debugToken: String? = if(!debug) null else {
-            val dt = abs(random.nextLong()).toString(16)
-            wsClientData.debugTokens[dt] = DebugInfo(javascript, "executing", Throwable())
-            dt
-        }
-        server2ClientMessage.debugToken = debugToken
-        wsClientData.handlers[server2ClientMessage.callbackId!!] = callback
-        wsClientData.send(server2ClientMessage)
+        wsClientData.handlers[funcCall.callbackId!!] = callback
```
Comment (sanity) [inline src/main/kotlin/kweb/Kweb.kt:203]: Might want to use `:? error(...)` to provide a more useful error message if callbackId is null.

File: src/main/kotlin/kweb/client/Server2ClientMessage.kt [hunk 6]
```diff
@@ -22,11 +24,21 @@ import kotlinx.serialization.json.JsonElement
 @SerialName("Server2ClientMessage")
 data class Server2ClientMessage(
         val yourId: String,
+        val functionCalls : List<FunctionCall>
+) {
+    constructor(yourId : String, functionCall : FunctionCall) : this(yourId, listOf(functionCall))
```
Comment (sanity) [inline src/main/kotlin/kweb/client/Server2ClientMessage.kt:29]: Is this constructor override necessary given the next one? Seems like the vararg should handle the case of a single FunctionCall.

File: src/main/kotlin/kweb/state/render.kt [hunk 7]
```diff
@@ -33,15 +36,48 @@ fun <T : Any?> ElementCreator<*>.render(
...
+                    block(value.value)
+                    if (renderState.get() == RENDERING_NO_PENDING_CHANGE) {
+                        renderState.set(NOT_RENDERING)
+                    }
+                }
+                null
+            }
+            if (outboundMessages != null) {
+                /*TODO to make this work I had to make WebBrowser.sessionId and Kweb.clientState public instead of private
+                  I think we may want to write a send() method in Kweb that we can call here, and set those vals back to private
```
Comment (sanity) [inline src/main/kotlin/kweb/state/render.kt:65]: Ah, I guess this answers my previous question.

File: src/main/kotlin/kweb/client/Server2ClientMessage.kt (hunk 6 above)
Comment (Derek52) [reply src/main/kotlin/kweb/client/Server2ClientMessage.kt:29]: I wrote both to have options. Currently, I'm not using the vararg constructor. The way we handle Server2ClientMessage creation, we really only need the secondary constructor that takes one item.

The only place in the code base where we generate a `Server2ClientMessage` with more than one `FunctionCall` is in `render()`, and I make a list manually there, so I use the default constructor.

So I was thinking of deleting the vararg constructor. You're right that we could use only the vararg constructor though.

File: src/main/kotlin/kweb/Kweb.kt (hunk 3 above)
Comment (Derek52) [reply src/main/kotlin/kweb/Kweb.kt:187]: That's a good thought. `JsFunction` only has 2 fields, and both of those are in `FuncCall`, so we should be able to just use `FuncCall`. That's an easy change for me to make.

File: src/main/kotlin/kweb/state/render.kt (hunk 7 above)
Comment (Derek52) [reply src/main/kotlin/kweb/state/render.kt:65]: Now that I think about it, I could also make public getters for these two fields and set the val's back to private. Should we go this route?

File: src/main/kotlin/kweb/client/Server2ClientMessage.kt (hunk 6 above)
Comment (sanity) [reply src/main/kotlin/kweb/client/Server2ClientMessage.kt:29]: I'm ok with either option.

File: src/main/kotlin/kweb/state/render.kt (hunk 7 above)
Comment (sanity) [reply src/main/kotlin/kweb/state/render.kt:65]: If they need to be publicly accessible then I don't think there would be any advantage to using separate getters, a `val` is essentially a getter.

**Pull request merged**